src/legacy/PhpCsFixer/Config.php

```diff
@@ -0,0 +1,11 @@
+<?php
+
+/**
+ * @copyright Copyright (C) Ibexa AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+declare(strict_types=1);
+
+use Ibexa\Platform\CodeStyle\PhpCsFixer\Config;
+
+class_exists(Config::class);
```

Comment:
I don't like this solution. Why not simple create   

```php
<?php

namespace EzSystems\EzPlatformCodeStyle\PhpCsFixer;


use Ibexa\Platform\CodeStyle\PhpCsFixer\Config as BaseConfig;

class Config extends BaseConfig {

}
```

instead of relying on class_exists implementation details?

Reply:
`class_exists` is only really needed to force PHP autoloader to load the new class. This causes both class names to appear in PHP memory (since the new one receives an alias there).

Using `class_alias` makes those classes in practice interchangeable. Every time an `EzSystems\EzPlatformCodeStyle\PhpCsFixer\Config` is requested, in reality an `Ibexa\Platform\CodeStyle\PhpCsFixer\Config` is used. Both type hints work as expected (i.e. you can typehint against either of those). 

Your proposed version would mean that in actuality two different classes would be present in PHP. While `EzSystems\EzPlatformCodeStyle\PhpCsFixer\Config` would be an instance of `Ibexa\Platform\CodeStyle\PhpCsFixer\Config`, the opposite would not be true, so you cannot typehint against the older class properly. You also cannot mark it as `final` even if you wanted to (in current PR's case, we could).

Reply:
> `class_exists` is only really needed to force PHP autoloader to load the new class. This causes both class names to appear in PHP memory (since the new one receives an alias there).
> 
> Note: Just using `Config::class` is not enough. PHP does not autoload on these. Class has to be actually used and checking against it's existence is the most explicit way of doing so.
> 
> Using `class_alias` makes those classes in practice interchangeable. Every time an `EzSystems\EzPlatformCodeStyle\PhpCsFixer\Config` is requested, in reality an `Ibexa\Platform\CodeStyle\PhpCsFixer\Config` is used. Both type hints work as expected (i.e. you can typehint against either of those).
> 
> Your proposed version would mean that in actuality two different classes would be present in PHP. While `EzSystems\EzPlatformCodeStyle\PhpCsFixer\Config` would be an instance of `Ibexa\Platform\CodeStyle\PhpCsFixer\Config`, the opposite would not be true, so you cannot typehint against the older class properly (newer class would not pass the type check). You also cannot mark it as `final` even if you wanted to (in current PR's case, we could).
> 
> EDIT: Also note that `class_exists` is used in a similar way to make PHP preload functionality load related classes.

This rather feels like utilization of side effects which might not last across PHP versions. I don't understand how it works.

Reply:
> This rather feels like utilization of side effects which might not last across PHP versions. I don't understand how it works.

This is exactly how this function is designed to work. It's supposed to force autoload to work unless second argument is passed as `false` (which disables the autoload attempt).

I find it hard to believe that this would change since it takes autoload into consideration by default. It would be a BC break in PHP, since it would change the default behavior.

Symfony uses it to force autoloading of classes required for `opcache.preload` functionality: https://github.com/symfony/symfony/blob/494ef421c554a78b38c6779c4b7deb9a20d89923/src/Symfony/Component/HttpKernel/HttpKernel.php#L35-L45

Reply:
`class_exists` (and family: `interface_exists`, `trait_exists`) is in practice the only function that I know of that will trigger autoloading (by default) without straight up using a class.
That is the reason it's used here. 

It is required due to the fact that return & argument type declarations **do not trigger autoload**. Because of this `class_alias` has to be put in the "new" class (otherwise old class would not be recognized when used as type). This means that the "old" class has to trigger autoloading, which is done via `class_exists`.

`class_alias` provides a low cost maintenance (no new class introduction) and allows for an "easy" upgrade path for users: both old and new FQCNs will work, so codebase can be adjusted using the same library version. Classes can be used interchangeably, providing a smooth upgrade.
Whether users upgrade all old references at once or do that in parts does not matter - both will continue to work normally.